오픈마켓 [STEP 1] 민쏜, 예톤

minsson opened the pull request:
@GREENOVER 

안녕하세요, 그린! 
민쏜과 예톤입니다! 😁

드디어 PR을 보냅니다..!!
이번 프로젝트는 워낙 생소한 개념이 많아, 프로젝트 초반에 공부하는 시간이 많이 들었던 것 같습니다.
아래에는 저희가 이번 프로젝트를 진행하며 고민했던 점을 작성해두었습니다.

조언 부탁드립니다! 감사합니다! 😁

## 🔎 고민한 점 

## 🥾 STEP 1

### 1. 네트워크 상황과 무관한 네트워킹 데이터 타입의 Unit Test

#### 현재의 코드 (테스트 불가능한 코드)
```swift
struct DataManager {
    static func performRequestToAPI(with request: String, completion: @escaping ((Data) -> Void)) {
        
    let task = URLSession.shared.dataTask(with: url) { }
    task.resume()
}
```

- 모든 화면에서 각 화면 별로 네트워크와 통신하며 데이터를 가져와야 한다고 생각했습니다.
- 이러한 관점에서 데이터와 통신하는 메서드는 공통적으로 쓰일 것이므로, `DataManager` 타입을 만들고 `타입 메서드`로 구현한다면 각 화면에서 `DataManager` 객체에 의존하지 않는 장점이 있다고 생각했습니다. (최근 의존성 주입에 대한 공부를 한 영향을 받아 각 객체 내부에서 인스턴스 생성을 피하고 싶다는 마음도 있었습니다.)
- 그런데 이 코드를 바탕으로 네트워크와 무관한 네트워킹 데이터 타입의 단위 테스트를 하려고 하자 문제가 발생했습니다.

#### 네트워크와 무관한 코드를 작성하는 과정과 문제
```swift
struct DataManager {
    
    private static let session: URLSession = URLSession.shared
    
    static func performRequestToAPI(with request: String, completion: @escaping ((Data) -> Void)) {
        
        let task = session.dataTask(with: url) {  }
        task.resume()
    }
}
```

- `야곰닷넷`의 `Unit Test` 강의에서 관련 내용을 공부해 구현하려고 했는데, 기존의 `타입 메서드`를 유지하기 위해서 `session` 변수 또한 `타입 프로퍼티`로 생성해주어야 했습니다.
- `타입 프로퍼티`의 경우 클래스의 `init` 메서드에 구현할 수 없었습니다.
- `init` 메서드에 구현할 수가 없으니 `Unit Test`에서 `StubURLSession` 등으로 내용을 교체할 수가 없었습니다.

#### 테스트 가능한 코드 vs 불가능한 코드
- 결론적으로, 저희가 작성한 코드는 `테스트 불가능한 코드`라고 생각했습니다. 
- `테스트 가능한 코드`로 만드려면 `DataManager`의 `타입 메서드`를 일반적인 `인스턴스 메서드`로 변경한 후, 뷰컨트롤러에는 `DataManager`의 인스턴스만 생성해주면 됩니다.
- 테스트가 필수라면 아직 수정이 용이한 현 시점에 변경하는 것이 가장 경제적이라고 생각했습니다.
- 하지만 단순히 `제공된 Mock 데이터를 활용해 모델이 네트워크 통신의 결과 데이터를 문제 없이 표현하고 활용할 수 있는지` 테스트하기 위해 저희가 효율적이라고 생각하는 구조를 변경해야 하는 건지 의문이 들었습니다 
    - 특히, 이런 테스트는 네트워킹과는 별도로 `parse` 메서드의 정상 작동 여부를 확인하는 것으로 충분할 수 있다는 생각도 들었습니다.
    - 물론, `타입 메서드`를 이런 방식으로 활용하는 게 정말로 좋은 구조인지는 모르겠습니다. 이 부분도 궁금합니다. 
    - 사실 이런 구조를 유지하는 것의 효용성과 유닛 테스트의 효용성을 대략적이나마 비교할 수 있다면 의사결정이 용이했을 거라는 생각이 듭니다. 더 많은 공부와 시행착오가 필요할 것 같습니다.


---

### 2. 각 화면별 모델의 구성
#### 처음의 생각 (효율성 및 최적화 측면)
- 서버에서 데이터를 받아오면 `상품명`, `상품번호` 등 모든 화면에서 공용으로 사용되는 데이터가 많습니다.
- 처음에는 각 화면에서 데이터를 가져오는 게 비효율적이니, 한번 서버로부터 가져온 데이터는 일정 한도 내에서 메모리에 보관해놓고, 모든 화면에서 공유하면 좋겠다는 생각이 들었습니다.
- 테이블뷰에서 위 아래 10개씩만큼의 데이터를 미리 받아놓는 등, 어떤 데이터를 어떤 기준으로 받아놓을지 결정해야 할텐데, 이게 유의미한 고민인지 의문이 들었습니다.
- 그럼에도 구현할 수만 있다면 앱의 속도가 더욱 빨라질 수 있을 것이라고 생각했습니다.


#### 현재의 생각 (구현 용이성, 유지보수 용이성 측면)
- 위의 고민을 하며 다른 앱에서는 어떤 식으로 구현해놓았는지 궁금했고, 가장 자료가 많을 것 같은 당근마켓 앱을 살펴보았습니다.
- 그동안 당근마켓 앱에서는 아무리 스크롤을 내려도 다른 앱들과 달리 데이터 로딩에 버벅임이 없었던 경험이 있어, 저희가 위에서 서술한 내용으로 구현되어 있을 수도 있다고 생각했습니다.
- 이를 위해서 중고 물품이 테이블뷰로 나오는 화면에 들어가 일정 시간을 기다린 후, 인터넷 접속을 해제하고 물품 중 하나를 탭해 상세 글 화면으로 이동해보았습니다. 예상과 달리 네트워크 연결 오류로 다음 화면의 내용이 표시되지 않았습니다.
- 실제로 네트워크에 접속 중인 상태에서도 화면 이동시 물품 이미지 사진 로딩에 아주 약간의 지연이 발생하며, 로딩되는 동안 회색화면으로 보이는 것을 알 수 있었습니다.
- 이런 대규모 서비스도 저희가 처음 한 생각대로 구현해놓지 않는다면, 안 하는 이유가 있을 거라고 생각했습니다.
    - 기술적으로 불가능하다.
    - 기술적으로 가능하지만, 이를 통해 유저가 느낄 효용에 비해 유지보수 등의 난이도가 급격하게 상승한다
- 따라서 향후 Step에서 여러 화면을 구현할 때 각 화면 별로 필요한 데이터를 비동기로 받아오는 것이 좋겠다는 생각이 들었습니다.


---

### 3. json 데이터 파싱
#### json 데이터를 받아오는 네이밍 변경
- Mock 데이터로 사용한 "products.json"에서는 개별 항목을 `page`라고 네이밍하고 있습니다.
- 저희는 page가 상품 전체를 보여주는 이름이 더 적절해보여, 해당 데이터를 받는 타입을 `ItemListPage`라고 정의했습니다.
- 기존 `page`는 `Item`으로 변경했습니다.

#### json 데이터를 받아오는 구조
- 각 아이템의 `name`과 `currency`를 `String`으로 받을지, `Enum`으로 받을지 고민했습니다.
- `currency`의 경우 화폐의 종류가 한정되어 있어 `enum`으로 바꾸면 더 편하고 효율적이라고 생각합니다.
- `name`의 경우 오타 등을 방지하기 위해 `상품 등록` 시에 입력받은 이름을 `enum`으로 받아 서버와 앱 전체에서 공유하면 좋겠다는 생각이 들었습니다. 
    - 하지만 데이터를 서버에서 받아오다 보니 저희가 관여할 수 있는 영역이 아닐 것 같다는 생각이 들었습니다. 앱 개발자가 이러한 영역에도 관여할 수 있는지, 실무에서는 어떤지 궁금합니다.

---

### 4. URLSessionDataTask - init() deprecated 
- 야곰닷넷의 unit test강의를 보며 테스트 코드를 작성하다가 `init` 부분에서 워닝을 발견했습니다.

```swift
class StubURLSessionDataTask: URLSessionDataTask {
    var dummyData: DummyData?

    // 여기
    init(dummy: DummyData?, completionHandler: DataTaskCompletionHandler?) {
        self.dummyData = dummy
        self.dummyData?.completionHandler = completionHandler
    }

    override func resume() {
        dummyData?.completion()
    }
}
```
에러의 내용은 아래와 같습니다.
> 'init()' was deprecated in iOS 13.0: Please use -[NSURLSession dataTaskWithRequest:] or other NSURLSession methods to create instances

- 이 경고를 해결하기 위해서 여기저기 검색해보다가 `URLProtocol`을 상속받은 `MockURLProtocol`을 사용하여 init을 사용하지 않는 방법과, 혹은 `URLSessionDataTaskProtocol`을 만들어서 `URLSessionDataTask`를 `extension`해주어 init을 사용하지 않는 방법 두 가지 방법에 대해 알게되었습니다.
- 하지만 두 가지 방법 모두 이해하기가..조금 어려워서 적용하지 못했습니다. 저희가 step 1의 Mock 데이터 테스트를 추후에 구현하게 된다면 이 경고를 해결하고 싶은데 좋은 방법이 있을지 그린에게 여쭤보고 싶습니다. 


---

### 5. 구조체 타입 내부 Enum
#### Enum을 구조체 타입 내부에 넣어줄 시 장단점

- 지금까지 프로젝트할 때에는 `Name Space`를 제거하기 위해서 `Name Space`를 `Enum` 타입으로 만들어서 전역에서 사용가능하도록 해주었습니다. 
- 그러나 이번 프로젝트에서는 민쏜의 의견에 따라 클래스 타입 내부에서만 사용되는 Name Space들을 정의해주기 위해서 타입 내부에 enum을 선언해준 뒤 `private`으로 접근 제한을 걸어주었습니다.
- 이렇게 타입 내부에 필요한 enum들만 생성해주고, 나중에 뷰컨트롤러가 더 많이 생길 때 공통되는 enum을 전역으로 한번에 빼주려고 생각해서 이렇게 코드를 작성했는데, 전역으로 enum을 선언하는 것과 타입 내부에 enum을 선언하는 것의 장단점이 무엇인지 궁금합니다.

---

잘 부탁드립니다!
감사합니다! 😁

yeeton37 committed:
feat: "ItemListPage" 상품 리스트 조회 페이지를 위한 구조체 타입 정의

- 오픈마켓 API로부터 파싱한 JSON 데이터와 매핑할 모델(ItemListPage) 설계
- json 데이터로 받아온 key의 이름을 직관적으로 표현하고자 pages -> items로 변경
- ItemListPage 내부에서만 쓰이는 Item 타입을 구조체 내부에 중첩 타입으로 선언

minsson committed:
feat: "DataManager" json 파일 디코딩을 위한 타입 메서드 구현

- "ItemListPage" swift 네이밍 컨벤션에 적절치 않은 일부 단어 수정

minsson committed:
feat: "DataManager" 오픈마켓 API에 데이터 요청하는 타입 메서드 구현

- 모든 뷰컨트롤러에서 공통적으로 데이터를 요청 및 수신해야 하므로 타입 메서드로 구현

minsson committed:
refactor: "DataManager" 데이터에셋 생성 및 디코딩하던 메서드 기능 분리

- 기존 parse(_:into:)의 기능 일부를 makeDataFrom(fileName:)으로 분리
- 네트워킹하여 받아온 data와 프로젝트 내부의 Mock 데이터 asset 모두에 범용적으로 사용하기 위함

yeeton37 committed:
style: 폴더 구조 정리

- Resources, Sources(Models, Views, Controllers)로 폴더 분리

yeeton37 committed:
refactor: "ItemListPageViewController"  매직스트링 제거

- queryString 프로퍼티를 생성할 때 필요한 private enum 선언
  (QueryValue, QueryKey, Path, QueryCharacter)

yeeton37 committed:
refactor: "DataManager" 데이터 요청 타입 메서드의 일부 기능 분리

- performRequestToAPI(with:completion:) 내부 response를 확인하는 부분을 isValidResponse(_:)로 분리

minsson committed:
test: "DataManagerTests" 데이터 디코딩 관련 메서드 유닛 테스트

- DataManager.makeDataFrom(fileName:) 메서드 테스트
- DataManager.parse(_:into:) 메서드 테스트

minsson committed:
feat: "DataManager" 네트워킹 중 에러처리를 위한 Result 타입 적용

- performRequestToAPI(with: completion:) 메서드에 Result 타입 적용
- 에러처리를 위한 APIError 타입 구현
- 해당 메서드 관련된 부분 파라미터 구성 등 수정

minsson committed:
style: 코드 전체에 가독성 개선을 위한 MARK 주석 작성

- 불필요한 개행 수정
- 불필요한 import Foundation 제거

minsson committed:
docs: README.md Step 1 관련 내용 업데이트

yeeton37 committed:
refactor: 불필요한 개행 제거

GREENOVER commented on OpenMarket/DataManagerTests/DataManagerTests.swift:
```diff
@@ -0,0 +1,52 @@
+@testable import OpenMarket
+
+class DataManagerTests: XCTestCase {
+    var mockData: NSDataAsset!
```
force Unwrapping 외 다른 방법은 없을까요?🙋🏻
물론 테스트 코드이니 런타임 시 앱 크래쉬를 유발하지 않겠지만 강제 언래핑 습관을 지양하면 어떨까합니다.
혹여 꼭 필요하거나 들어가는게 더 적절하다고 판단되는 강제 언래핑도 있지만 이 경우에는 어떨지 궁금해요~!
그런데 현재 mockData 변수는 사용되지 않고 있는데 사용될 여지가 있는건가요~?

minsson replied:
`var mockData: NSDataAsset?` 과 같이 선언하고, 사용시에 `옵셔널 바인딩`을 하면 될 것 같습니다! 그리고 현재 mockData 변수는 사용되지 않으므로 제거해주었습니다.

GREENOVER commented on OpenMarket/DataManagerTests/DataManagerTests.swift:
```diff
@@ -0,0 +1,52 @@
+class DataManagerTests: XCTestCase {
+    var mockData: NSDataAsset!
+
+    override func setUpWithError() throws {
+        try super.setUpWithError()
+    }
+
+    override func tearDownWithError() throws {
+        try super.tearDownWithError()
+    }
```
여기서 추후 mockData를 사용하게 된다면 초기값을 주거나 해제하는등의 작업이 필요할거라 예상되네요🤔

GREENOVER commented on OpenMarket/DataManagerTests/DataManagerTests.swift:
```diff
@@ -0,0 +1,52 @@
+        try super.tearDownWithError()
+    }
+
+    func test_makeDataFrom메서드_json파일명을_넣을시_옵셔널Data타입으로_반환한다() {
```
두 테스트 코드 외에 조금 더 많은 테스트 코드를 작성하여 검증을 해보면 어떨까요~?
상품 정보에 대해서도 확실히 테스트를 해보면 좋을거같아요..!
더 나아가서 현재는 GREEN 케이스의 테스트이니 RED 즉 실패하는 테스트 코드도 구성해본다면 TDD의 기본 요구를 따를거라 생각해요🙋🏻

yeeton37 replied:
앗 저희가 이해하기로는 TDD를 작성할 때 RED, 즉 실패하는 케이스를 작성한다고 해도 추후에 최소한의 실행가능한 코드를 구현해주고나면 RED였던 코드도 GREEN으로 변하는 것이라고 생각했는데, 저희가 최종 PR을 보낼 때는 모든 테스트 코드가 정상 실행되는 상태로 보내지게 되는것이라서.. 혹시 실패 가능한 테스트 코드를 남겨두는 방법이 따로 있을까요?

GREENOVER replied:
아하! 아 제가 간과했던 부분이네요!
두분이 이해하시고 말씀하신게 맞다고 생각합니다👍
제가 너무 코드에 보이지 않아서 단편적으로 생각했네요🥲
혼선을 드려 죄송합니다..!
킵고잉 해주시져ㅎㅎ

GREENOVER commented on OpenMarket/DataManagerTests/DataManagerTests.swift:
```diff
@@ -0,0 +1,52 @@
+            return
+        }
+        
+        itemListPage = parsedData
```
parse 메서드 구현에서 코멘트를 남겼지만 parse가 target 인자를 받아서 해당 target 인자의 type을 추출해서 targetType으로 사용하다보니 현재 스텝이 하나 더 는것처럼 느껴져요🤔
itemListPage 옵셔널 변수를 선언하고 이에 다시 parsedData를 할당하는것처럼요.
이 부분의 코드를 하나로 깔끔히 할 수도 있을거라 생각해요.
물론 parse 구현을 고치면요..!🙋🏻

yeeton37 replied:
그러네요..! 기존에는 parse의 매개변수로 인스턴스를 받아와서 메서드 내부에서 타입을 추론 후 decode 메서드 안에 넣어주었는데,
그린의 코멘트를 보고 decode할 타입 자체를 매개변수로 받아올 수 있다는 것을 알게되었습니다.ㅎㅎ
이에 따라 매개변수로 타입 자체를 받아줄 수 있도록 제너릭 타입(T)에서 제너릭의 metaType(T.Type)으로 변경해주고 나니 parse 메서드를 호출하는 부분도 훨씬 간편하게 변경해줄 수 있게 되었네요! 감사합니다.😁

#### 구현부
```swift
 static func parse<T: Decodable>(_ data: Data, into type: T.Type) -> T? {
        let jsonDecoder: JSONDecoder = JSONDecoder()
                
        do {
            let decodedData = try jsonDecoder.decode(T.self, from: data)
            return decodedData
        } catch {
            return nil
        }
    }
```
#### 변경 전 호출부
```swift
func fetchParsedData(basedOn result: Result<Data, NetworkingError>) {
        switch result {
        case .success(let data):
            guard let parsedData = DataManager.parse(data, into: self.itemListPage) else {
                return
            }
            
            self.itemListPage = parsedData
        case .failure(let error):
            print(error)
        }

```
#### 변경 후 호출부
```swift
private extension ItemListPageViewController {
    func fetchParsedData(basedOn result: Result<Data, NetworkingError>) {
        switch result {
        case .success(let data):
            itemListPage = DataManager.parse(data, into: ItemListPage.self)
        case .failure(let error):
            print(error)
        }
    }
}
```

GREENOVER commented on OpenMarket/OpenMarket.xcodeproj/project.pbxproj:
```diff
@@ -49,24 +123,37 @@
-				C70FB10325BEF61D00C9924E /* Assets.xcassets */,
-				C70FB10525BEF61D00C9924E /* LaunchScreen.storyboard */,
 				C70FB10825BEF61D00C9924E /* Info.plist */,
+				4252743228810FC700A63D7F /* Resources */,
+				425274352881103B00A63D7F /* Sources */,
```
오호~!
이렇게 프로젝트 내부에서 Resource그룹과 Source그룹으로 분리해둔다면 어떤 이점이 있을까요?
또 어떤 기준으로 그룹핑해주셨을지도 궁금합니다.
추후 프로젝트 모듈화등도 고려될 수 있는걸까요?🙌

minsson replied:
### 그룹핑 기준
평소에 프로젝트를 진행하면서 어떤 파일을 어떤 폴더에 넣는 게 좋을지 고민이었고, 일단은 저만의 기준이 잡히기 전까지 다른 개발자 분들의 폴더 구조를 벤치마킹하는 것도 방법이라고 생각했습니다.

따라서 [File Finder · devxoul/RxTodo](https://github.com/devxoul/RxTodo/find/master)의 폴더구조를 따라해보았습니다만, 그룹핑에 대한 정확한 기준은 정하지 못한 상태입니다. 더 큰 규모의 프로젝트를 수행하면서 많은 파일을 관리해야 할 때 시행착오를 겪으며 기준을 정할 수 있을 것 같다고 생각합니다.

###  Resource그룹과 Source그룹으로 분리해둘 때의 이점
Resource 그룹에는 앱을 만들 때 리소스로 쓰일 Assets가 들어가고, Sources에는 현재 저희가 쓰고 있는 Model, View, Controller는 아니면서, 앱의 구동에는 필요한 소스코드가 들어간다고 생각했습니다.

이렇게 분리할 때의 이점은 아직 확실히 와닿지가 않습니다. 다만 프로젝트 규모가 커졌을 때 Resources 폴더의 파일은 필요에 따라 유동적으로 교체되고, 삭제될 수 있는 반면, Sources 그룹의 파일들은 필수적으로 유지될 파일이라고 생각합니다. (아마 아래에 작성한 프로젝트 모듈화와 관련이 있을 것 같습니다)

### 프로젝트 모듈화
이 부분은 지식이 부족해 아직 정확히 어떤 개념인지 잘 이해되지 않습니다. 프로젝트가 커질수록 각 파일의 기능 분리를 명확히 해서 코드 재사용성을 증가시키고, 역할 및 책임을 나누어 유지보수를 원활하게 만들 필요가 생겨 프로젝트 모듈화를 한다고만 이해했습니다. 이와 관련 Tuist라는 프레임워크도 사용하고, 에어비앤비에서는 bucks라는 걸 사용한다고 하는데, 이런 부분은 더 공부해야겠다는 생각이 들었습니다.

한번 공부해놓으면 작은 프로젝트가 점점 규모가 커질 때 겪을 시행착오를 사전에 대비할 수 있을 것 같습니다.

minsson replied:
### 그룹핑 기준

평소에 프로젝트를 진행하면서 어떤 파일을 어떤 폴더에 넣는 게 좋을지 고민이었고, 일단은 저만의 기준이 잡히기 전까지 다른 개발자 분들의 폴더 구조를 벤치마킹하는 것도 방법이라고 생각했습니다.

따라서 [File Finder · devxoul/RxTodo](https://github.com/devxoul/RxTodo/find/master)의 폴더구조를 따라해보았습니다만, 그룹핑에 대한 정확한 기준은 정하지 못한 상태입니다. 더 큰 규모의 프로젝트를 수행하면서 많은 파일을 관리해야 할 때 시행착오를 겪으며 기준을 정할 수 있을 것 같다고 생각합니다.

### Resource그룹과 Source그룹으로 분리해둘 때의 이점

Resource 그룹에는 앱을 만들 때 리소스로 쓰일 Assets가 들어가고, Sources에는 `코드`가 들어간다고 생각했습니다.

이렇게 분리할 때의 이점은 아직 느껴보지 못해서 잘 모르겠습니다. 현재로서는 계속 수정하고 신경써야할 부분은 Sources에, 그렇지 않은 부분은 Resource에 있어 보기에 깔끔하다는 느낌 정도입니다. 

그린이 생각하시기에는 어떤 이점이 있나요? 그리고 프로젝트 모듈화와는 어떤 관계가 있나요? 😲

### 프로젝트 모듈화

이 부분은 지식이 부족해 아직 정확히 어떤 개념인지 잘 이해되지 않습니다. 프로젝트가 커질수록 각 파일의 기능 분리를 명확히 해서 코드 재사용성을 증가시키고, 역할 및 책임을 나누어 유지보수를 원활하게 만들 필요가 생겨 프로젝트 모듈화를 한다고만 이해했습니다. 이와 관련 Tuist라는 프레임워크도 사용하고, 에어비앤비에서는 bucks라는 걸 사용한다고 하는데, 이런 부분은 더 공부해야겠다는 생각이 들었습니다.

한번 공부해놓으면 작은 프로젝트가 점점 규모가 커질 때 겪을 시행착오를 사전에 대비할 수 있을 것 같습니다.

GREENOVER commented on OpenMarket/DataManagerTests/DataManagerTests.swift:
```diff
@@ -0,0 +1,52 @@
+//  Created by minsson, yeton on 2022/07/15.
+//
+
+import XCTest
+@testable import OpenMarket
```
현재 테스트 타겟의 배포 타겟이 15.5로 설정되있고 OpenMarket의 배포 타겟은 13.2로 되있어서 15.5 미만 즉 13.2 부터 15.4까지의 타겟에서 테스트가 불가합니다🥲
테스트 타겟의 배포 타겟을 같이 맞춰주면 어떨까요~?🙋🏻

yeeton37 replied:
감사합니다. 테스트 타겟의 배포 타겟을 DataManagerTests>Build Settings>iOS Development Target로 들어가서 iOS 13.2 로 변경해주었습니다!

GREENOVER commented on OpenMarket/DataManagerTests/DataManagerTests.swift:
```diff
@@ -0,0 +1,52 @@
+        XCTAssertTrue(type(of: result) == Data?.self)
+    }
+    
+    func test_mockData를_parse하면_첫번째_아이템_id는_20이다() {
```
이건 정말 스타일의 차이지만 테스트 코드 메서드 네이밍을 한글 / 한글+영어 / 영어 이 세가지 조합으로 이뤄질 수 있을것 같은데 여러분들은 해당 방식을 선택하신 이유가 궁금합니다🙋🏻
그리고 또 현재 방식과 제시된 방식들의 장,단점은 어떤게 있을까요?

yeeton37 replied:
1. 한글로 네이밍할 경우: 
- 장점: 한글이다보니 직관적이고, 테스트 목적과 내용을 명확하게 이해할 수 있습니다.
- 단점: 한글이니까 외국인은 못알아볼 것같아요.
2. 영어로 네이밍할 경우:
- 장점: 범용적으로 사용될 수 있습니다.
- 단점: 이해하는 데 시간이 걸릴 수도 있습니다. 한국인 입장에서 직관적이지 못합니다.

3. 한글+영어로 네이밍할 경우:
저희는 이 방식으로 테스트 메서드를 네이밍해주었습니다.
테스트하는 메서드의 이름이나 프로퍼티 등은 한글로 작성하는 것보다 영어 그대로 작성하는 것이 코드를 볼 때 한눈에 알아보기에 편할 것 같아서 영어로 적어주었고, 나머지는 테스트 코드의 메서드를 읽을 때 직관적으로 빠르게 해석이 될 수 있도록 한글로 작성해주었습니다.!

GREENOVER commented on OpenMarket/DataManagerTests/DataManagerTests.swift:
```diff
@@ -0,0 +1,52 @@
+        
+        // when
+        guard let parsedData = DataManager.parse(data, into: itemListPage) else {
+            return
```
실제 파싱의 값이 없거나 오류가 나 이 else문의 return을 타게 될때도 해당 테스트는 success를 배출할거라고 생각해요🙋🏻
간단하게 43번째 라인에 return을 넣고 expectation을 21로 바꾸고 테스트 해보면 해당 코멘트를 이해할 수 있을거라 생각합니다..!
즉 해당 테스트 코드의 메서드 시그니처죠 () -> () 자체를 return으로 만족했기에 더이상 아래 코드를 실행하지 않으며 그렇기에 XCTAssertEqual도 실행하지 않을것이며 최종적으로 무사 통과되었다고 보고할거에요🥲
그렇다면 어떻게 보완하면 좋을까요~?🙋🏻

minsson replied:
XCTAssert 테스트 메서드를 만나기 전, `return` 키워드를 만나면 무조건 성공하게 되는 경우의 수를 고려하지 못했네요..!! 😲
`return`키워드 전에 `XCTFail()` 메서드를 추가하여 해결했습니다!

말씀해주신 `parse(_:into:)` 수정 과정에서 guard let이 필요하지 않게 되었고, 아래와 같이 다른 메서드에 적용해보았어요.

```swift
func test_mockData를_parse하면_첫번째_아이템_id는_20이다() {
        // given
        let filename = "products"
        guard let data = NetworkManager.makeDataFrom(fileName: filename) else {
            XCTFail("JSON 데이터 파일명 및 형식 확인 필요")
            return
        }
        var itemListPage: ItemListPage?
        
        // when
        let parsedData = NetworkManager.parse(data, into: ItemListPage.self)
        itemListPage = parsedData
        
        let result = itemListPage?.items[0].id
        let expectation = 20
        
        // then
        XCTAssertEqual(result, expectation)
    }
```

GREENOVER commented on OpenMarket/OpenMarket/Sources/Models/Errors/APIError.swift:
```diff
@@ -0,0 +1,16 @@
+
+// MARK: - Enum
+
+enum APIError: Error {
```
에러를 아주 세분화시켜 열거해주신점 아주아주 좋네요👍

GREENOVER commented on OpenMarket/OpenMarket/Sources/Models/Errors/APIError.swift:
```diff
@@ -0,0 +1,16 @@
+//
+//  NetworkingError.swift
```
NetworkingError인지 APIError인지 통일시키면 더 좋을것 같아요🙌
추가로 해당 에러 파일 및 이 파일이 속한 Errors 그룹 자체를 Model에서 별도로 분리시켜 같은 계층구조를 가지게 하면 조금 더 적절하지 않을까란 생각이 있습니다🙋🏻
왜냐하면 이 에러는 모델에서만 국한되는것이 아니라는 생각이 들었어요.
네트워크 전반의 에러이니 네트워크 쪽에 묶여야되지 않을까..?란 생각이 들었습니다..!

yeeton37 replied:
네트워크 전반의 에러이니까 네이밍은 `NetworkingError`으로 통일시켜주었습니다!
그린의 말씀대로 error는 모델에서만 국한되는 것이 아니라 네트워크 자체의 에러를 나타내는 것이기 때문에 모델과 같은 뷰 계층구조를 가진 곳으로 Errors 폴더를 옮겨주었습니다.

GREENOVER commented on OpenMarket/OpenMarket/Sources/Models/Errors/APIError.swift:
```diff
@@ -0,0 +1,16 @@
+// MARK: - Enum
+
+enum APIError: Error {
+    case decodingJson
```
URL, API와 같이 JSON도 고유하게 카멜 케이스를 따르지 않고 전부 대문자로 표현하는게 조금 더 친숙하지 않을까해요🤔
저는 요렇게 표현하고 있는데요.
1. 프로퍼티 및 메서드의 처음에 시작함 -> `jsonDecoding`
2. 중간에 위치함 -> 'decodingJSON`
즉 정리하자면 JSON, json으로 사용하며 카멜케이스를 이런 경우는 지켜주고 있지 않습니다.
이것도 정말 정말 스타일의 차이긴 하지만 애플에서도 `JSONDecoder`와 같이 사용하는것과 많은 오픈 소스들이 위와 같이 사용함으로 URL, API와 같이 취급해주면 더 적절할것 같습니다🙋🏻

yeeton37 replied:
앗! 저희의 실수였네요 😅 앞으로 json이 프로퍼티, 메서드의 처음에 시작하는 것은 소문자 형태로, 중간에 위치하는 경우에는 대문자 형태로 써주어야겠단 걸 알게되었네요. 감사합니다! `decodingJSON`으로 네이밍 변경해주었습니다 ㅎㅎ

GREENOVER commented on OpenMarket/OpenMarket/Sources/Models/Errors/APIError.swift:
```diff
@@ -0,0 +1,16 @@
+// MARK: - Enum
+
+enum APIError: Error {
+    case decodingJson
+    case data
+    case client
+    case server
```
어떤 에러인지 조금 더 명확한 네이밍은 어떨까요~?🙋🏻
`decodingJSONFailed`와 같이 실패한 케이스인지 어떤건지 이 케이스명만 보고 유추할 수 있는게 좋지 않을까 생각해요...!

yeeton37 replied:
앗 그러네요. 😅 아래와 같이 조금 더 구체적으로 에러명을 변경해주었습니다!
```swift
enum NetworkingError: Error {
    case decodingJSONFailure
    case invalidURL
    case clientTransport
    case serverSideInvalidResponse
    case missingData
    case unknown
}
```

GREENOVER commented on OpenMarket/OpenMarket/Sources/Models/ItemListPage.swift:
```diff
@@ -0,0 +1,64 @@
+    
+    private enum CodingKeys: String, CodingKey {
+        case pageNumber = "page_no"
+        case itemsPerPage = "items_per_page"
+        case totalCount = "total_count"
+        case offset, limit
+        case items = "pages"
+        case lastPage = "last_page"
+        case hasNext = "has_next"
+        case hasPrevious = "has_prev"
```
만약 pageNumber, items와 같이 아예 다른 네이밍으로 바꿔주지 않고 스네이크 케이스를 단순 카멜 케이스로 디코딩 시 변환하고 싶다면 애플에서 제공하는 유용한 기능도 있습니다.
이걸 사용하는것은 사실 변수명을 바꾸지 못하기에 국한될 수 있지만 아~ 이런것도 있구나 참고하시면 지식+1이 되실것 같아 코멘트 드립니다..! (이미 학습하셨을수도 있지만..!🥲)
https://developer.apple.com/documentation/foundation/jsondecoder/keydecodingstrategy/convertfromsnakecase

yeeton37 replied:
오...!저희는 이름을 변경해준 부분이 있어서 이번 프로젝트에서는 적용하지 못하겠지만, 추후에 엄청 유용하게 사용해줄 수 있겠네요! 기록해두겠습니다 감사해요 😇

GREENOVER commented on OpenMarket/OpenMarket/Sources/Models/ItemListPage.swift:
```diff
@@ -0,0 +1,64 @@
+    
+    // MARK: - Properties
+    
+    let pageNumber, itemsPerPage, totalCount, offset: Int
```
이 4개 프로퍼티만 묶어주신 이유는 page라는 공통 관심사에 속하는 프로퍼티들이라고 보면 될까요~?🙋🏻
아주 좋네요ㅎㅎ

GREENOVER commented on OpenMarket/OpenMarket/Sources/Models/ItemListPage.swift:
```diff
@@ -0,0 +1,64 @@
+    
+    // MARK: - Nested Struct
+    
+    struct Item: Decodable {
```
인코딩이 현재 필요없기에 Decodable 프로토콜만 채택하신점 아주아주 좋네요👍
다만 현재 요구사항에는 나오지 않았지만 유추해본다면 아이템 즉 상품을 디코딩해서 받아올 수 있으면 이 프로젝트가 오픈마켓이니 상품을 등록할 수도 있어야겠죠!?
그럴때는 또 다른 requestDTO, 별도 모델 타입을 만들기보다 현재 만들어진 이 Item 타입을 활용할 가능성이 농후합니다.
그렇다면 추후를 고려해 Codable하면 어떨까 싶어요🤔
물론 추후에 고쳐도 무방합니다..!
현재로써 너무너무 잘해주셔서 한단계 더 나아가서 이 모델 타입이 쓰일 용도를 예측하는것까지 생각해보면 좋을것 같아 의견드립니다🙇🏻

minsson replied:
처음에는 `Codable`로 작성하려다 현재 단계에서는 `Encodable`이 필요하지 않으므로 `Decodable`로 수정해보았습니다. `추후 쓰일 수도 있으니 만들었다가 방치되는 코드`와 `확장성`에 대한 고민을 잠깐 했었어요. (이렇게 사소한 걸로 그런 고민을 할 필요는 없었을 수도 있지만요 😅)

그런데 그린 말씀을 듣고 나니 `Codable` 같은 경우는 쓰일 게 명백하기 때문에, 처음부터 `Codable`을 쓰는 게 나았겠다는 생각이 들었어요! 이 부분은 Step 2에서 수정해볼게요! 😁

GREENOVER replied:
오 명확한 근거를 지닌 코드였군요.
너무 좋습니다!👍👍👍

GREENOVER commented on OpenMarket/OpenMarket/Sources/Models/ItemListPage.swift:
```diff
@@ -0,0 +1,64 @@
+        case hasPrevious = "has_prev"
+    }
+    
+    // MARK: - Nested Struct
```
중첩된 구조로 ItemListPage에 Item 그 안에 Currency가 있는것의 이점이 어떤게 있을까요?🙋🏻
두가지 방법들이 더 있을텐데,
1. 현재 파일에서 중첩 구조가 아닌 각 타입들을 분리
2. 각 타입들을 다른 파일로 분리
현재 구현하신 스타일까지 총 3가지로 보여요.
그랬을때 추후의 유지보수 및 가독성까지 생각해본다면 어떤 방향이 적절하고 장단점이 있을까요?🙋🏻
이건 정답이 없는 문제입니다..! 여러분의 의견이 궁금해요ㅎㅎ

minsson replied:
말씀해주신대로 정답이 없는 문제라고 생각합니다.
다만 오픈마켓 프로젝트의 특성과 향후 프로젝트 진행 방향을 생각했을 때,
결론적으로 `Currency`를 다른 타입으로 분리하는 방안이 추후의 유지보수 및 가독성 등을 고려했을 때 가장 적절하다고 생각합니다.
근거는 아래와 같습니다.

(🟢: 최선 / 🟡: 차선 / 🔴: 차차선)

## 생각의 근거
### 1. 현재 구현한 스타일 🟡
- 중첩된 구조로 ItemListPage에 Item, 그 안에 Currency가 있는 상태
#### - 장점: 
- 현재의 코드에서 `Currency`는 `Item`에 속한 상태에서만 쓰입니다. 관계구조를 명확해지고, 다른 타입에 `Currency`를 보여주지 않음으로써 일종의 접근제어도 가능하다고 생각합니다. 
- 관계구조가 중요한 코드라면 가독성도 좋아보일 수 있다고 생각합니다.
- 이러한 장점으로는 3개의 방안 중 가장 유리합니다.
#### - 단점: 
- 상품 등록, 상품 상세보기 등 여러 페이지가 늘어나면 다른 뷰컨트롤러에서도 쓰일 것이 명확한데, 현재 구조를 유지한다면 모든 뷰컨트롤러에 해당 enum을 만들어줘야 해, 코드가 중복됩니다.
- 신규 국가에 앱을 론칭하는 등 화폐를 수정해야 하는 경우, 모든 뷰컨트롤러 파일에 들어가서 수정해줘야 해, 유지보수가 어려워집니다.
- 이러한 단점으로는 3개의 방안 중 가장 불리합니다.

### 2. 각 타입들을 다른 파일로 분리 🟢
#### - 장점:  
- 현재까지 밝혀진 프로젝트 요구사항과 앞으로의 확장성을 생각하면 뷰컨트롤러가 여러 개가 될 것 같습니다. 여러 뷰컨트롤러에서 접근하기 좋습니다.
- 3개의 방안 중 유지보수 측면에서 가장 유리합니다.
- 다양한 타입에서 접근한다고 생각한다면, 가독성 측면에서도 유리합니다.

#### - 단점: 
- 타입 간의 관계구조에 대한 명확도가 세 방안 중 가장 떨어집니다. 
(다만, 프로젝트 구조상 여러 뷰컨트롤러에서 활용할 타입이므로, 오히려 아무 타입에도 속하지 않는다는 점이 장점이 됩니다.)

### 3. 현재 파일 내에서 타입을 분리 🔴
#### - 장점: 
- 1번과 같이 관계구조를 보여줄 수 있으면서도 다른 뷰컨트롤러에서 접근할 수 있습니다.

#### - 단점:
- 이 방안은 좀 애매해보입니다. 관계구조도 표현할 수 있으면서 다른 타입도 접근할 수 있어 좋아보이지만, 오히려 개발자에게 혼동을 줄 수 있어 유지보수 측면에서도 1번보다 안 좋을 수 있다고 생각합니다.

GREENOVER commented on README.md:
```diff
@@ -1,7 +1,163 @@
-## iOS 커리어 스타터 캠프
+# 오픈 마켓
```
이렇게 완성도 있는 스텝1 리드미는 처음인것 같네요..!
금요일 리드미 스터디에서 잘 하시겠지만 추후 리팩토링을 통해 트러블 슈팅과 해당 파일들의 구조를 UML로 작성하신게 있다면 같이 올리면 금상첨화일것 같습니다👍👍👍
리드미 화이팅!

minsson replied:
미리미리 기록해놔야 나중에 잊어버리지 않을 것 같아서 작성해보았습니다! Step 2를 진행하면서 UML도 작성해 추가해보겠습니다! 감사합니다! 😁

GREENOVER commented on OpenMarket/OpenMarket/Sources/Models/DataManager.swift:
```diff
@@ -0,0 +1,79 @@
+
+import UIKit
+
+struct DataManager {
```
**<<요 코멘트는 API 네트워크 통신까지 해당 구조체에서 해줄때를 가정한 코멘트임으로 그게 아니라면 가볍게 스킵해도 됩니다!>**>

DataManager의 네이밍이 조금 다소 한정적일 수 있지 않을까란 생각이 조금 들었어요🙋🏻
실제 해당 타입의 메서드 구현들을 보면 여기서 Data는 사실상 실제로 사용될 Item 및 ItemListPage와 같은 데이터가 되는것 같은데 바로 이어서 스텝으로 있을 상품조회/등록/삭제 시 실제 responseDTO에 이러한 Item 데이터를 넘겨주지 않는 경우도 꽤 있습니다.
(해당 야곰 오픈마켓 프로젝트에서는 넘겨주더라도 다른 프로젝트에서는 안넘겨줘도 사실 상관없기에..)
그렇기에 조금 더 추상적으로 나아가서 이 구조체가 하는 역할은 제가 볼때 네트워크 통신을 전반적으로 다 담당해주는 친구라고 느껴졌어요.
만약 그게 맞다면 NetworkManager와 같은 포괄된 범위의 네이밍은 어떨까요~?🙋🏻
향후 이 구조체안에 실제 RESTful API를 통해 GET, POST, DELETE 모두 다 정의 및 구현될거라고 생각해서요!
그게 아니고 이 친구는 정말 이 역할만 하고 API 네트워크 통신은 또 다른쪽에서 하는것이라면 별도 분리하는게 맞으니 여러분들의 의도를 공유해주시면 감사하겠습니다🙇🏻

minsson replied:
해당 타입은 말씀해주신 것처럼 API 네트워크 통신하는 역할과, 데이터를 저희가 지정한 모델 타입으로 디코딩하는 역할 모두를 수행하고 있어요. 

따라서 말씀해주신 것처럼 `NetworkManager`라는 이름이 더 적절하게 느껴져 수정했습니다! 😁

다만, `바로 이어서 스텝으로 있을 상품조회/등록/삭제 시 실제 responseDTO에 이러한 Item 데이터를 넘겨주지 않는 경우도 꽤 있습니다.
(해당 야곰 오픈마켓 프로젝트에서는 넘겨주더라도 다른 프로젝트에서는 안넘겨줘도 사실 상관없기에..)` 라는 말씀이 잘 이해가 되지 않았어요.

> `DTO`는 아래와 같이 이해했어요!
> - 공개 변수만 있고 함수가 없는 구조체
> - DB나 API에서 저장된 가공되지 않은 정보를 애플리케이션 코드 안에서 처음으로 객체로 변환해줄 때 사용하는 구조체

그런데 `Response DTO`와, `Item 데이터를 넘겨주지 않는 경우`가 특히 잘 이해되지 않아요..! 혹시 조금만 더 자세히 설명해주실 수 있나요? 🥲

GREENOVER replied:
아아 제가 설명이 조금 애매했네요!
제가 여기서 말한 ResponseDTO는 조금 더 넓은 범위의 큰 개념이라고 생각하면 좋을것 같아요.
즉 예를들어 상품 수정, 등록, 삭제등을 request해주고 이에 따라 각 response가 있을텐데
여기서 기본적으로 RESTful API에서는 response로 API DTO를 뭘 줄지 설계하기 나름이겠지만
statusCode (200 등 상태값), 필요에 따라 message(에러 및 성공 시 메시지) 그리고 data (해당 프로젝트로 적용한다면 Item 이 될것 같아요) 로 소통을 해주기 마련입니다.
여기서 data라는것은 실제 상품 삭제시를 예를들면 삭제를 성공했다는 response만 있으면 되지 사실 어떤 상품을 삭제했는지 그 상품 정보를 response에서 data로 받을 필요는 없을때도 많아요..!
근데 현재 야곰 API 명세는 우선 넘겨주긴 하니까 받아야겠지만요..!?
그랬을때를 가정해서 이 data로 넘어오는것은 매번 다 다를텐데 그에 따라 responseDTO를 각각 작성하는것보다 아래와 같은 공통된 responseDTO를 작성해주면 어떨까란 의미에서 조금 스텝2에서 고민해볼 거리를 미리 말한 경향이 있습니다..!
```swift
public enum CommonResponseDto {
  public struct ExistData<T: Decodable>: Decodable {
    public var statusCode: Int
    public var message: String
    public var data: T?
...
  }

  public struct NotExistData: Decodable {
    public var statusCode: Int
    public var message: String
    ...
  }
}
```
여기서 본다면 이 DTO를 통해 다 해줄 수 있도록 구성했어요.
즉 data라는 것이 현재 정의하신 Model 타입으로 제네릭하게 들어갈거라 생각합니다.
음.. 조금 말로 표현하기가 어렵네요 ㅠ
이해가 안가신다면 그냥 아~ 정도만 하고 스텝2에서 겪은 다음 같이 회의실에서 얘기를 나눠보면 좋을것 같습니다🙌

GREENOVER commented on OpenMarket/OpenMarket/Sources/Models/DataManager.swift:
```diff
@@ -0,0 +1,79 @@
+    
+    // MARK: - Properties
+    
+    private static let openMarketHostAPI = "https://market-training.yagom-academy.kr"
```
추후 host가 여러개가 되면 여기서 모두 선언해줘야할것 같은데
별도 네임스페이스로 관리하면 어떨까요?🙋🏻

yeeton37 replied:
넵! 아래와 같이 enum으로 host API를 구현해주었습니다.
```swift
enum HostAPI {
    case openMarket
    
    var url: String {
        switch self {
        case .openMarket:
            return "https://market-training.yagom-academy.kr"
        }
    }
}
```

GREENOVER commented on OpenMarket/OpenMarket/Sources/Models/DataManager.swift:
```diff
@@ -0,0 +1,79 @@
+    
+    // MARK: - Static Actions
+    
+    static func performRequestToAPI(with request: String, completion: @escaping (Result<Data, APIError>) -> Void) {
```
이것도 정말 정말 스타일의 차이이긴한데 저는 아래와 같이 씁니다..!
```swift
performRequestToAPI(
  with request: String,
  completion: @escaping (Result<Data, APIError>) -> Void
) {
  ...
}
```
현재처럼 메서드 파라미터 시그니쳐가 길어지거나 하여 가로가 길어지는것이 생길때 적절한 개행을 해주고 있어요.
왜냐하면 개발자들의 디바이스는 모두 다를것이고 이렇게 가로가 늘어나게되면 사실 코드를 볼때 가독성을 해치는 경우가 많더라구요.
그렇기에 가로보다는 세로가 긴것이 원치않는 길이때문에 줄바꿈을 일으키지 않음으로 가독성에 개인적으로 좋다고 생각합니다..!
오픈 소스들도 시간날때 보시면 어떤 스타일을 채택하고 있는지 공부가 될거라 생각해요ㅎㅎ
저는 좀 극단적인 편이고.. (파라미터가 2개 이상 되면 무조건 위처럼 개행을 해주는 컨벤션을 가집니다!)
여러분들에게 맞는 스타일을 선택하면 될거에요!
정말 처음에도 말했듯 스타일의 차이이니 아 이런것도 있구나~ 하고 넘어가도 좋습니다👍
스타일쉐어가 코드 컨벤션 정리를 아주 잘 해두었다고 개인적으로 생각해 공유드립니다.
https://github.com/StyleShare/swift-style-guide#들여쓰기-및-띄어쓰기

minsson replied:
말씀하신 것처럼 취향의 차이라고 생각해요 😁
사실 이 부분은 저희도 나눌까말까 고민이 많았어요...!

처음에는 너무 길어서 나누고 싶었고, 나눠보니 보기가 좋았는데, `이걸 나누면 파라미터가 2개 이상인 다른 모든 메서드에도 적용해야 하지 않나?`라는 생각이 들었어요. 통일성 측면에서요!

하지만 두 명 모두 `파라미터가 2개 이상이어도, 길이가 짧다면 한 줄로 쓰는 게 더 보기 좋다`라는 점에 동의했고, 아쉽지만 통일성 측면에서 `performRequestToAPI(with:completion:) 메서드도 한 줄로 쓰기로 결정했어요.

그렇지만 너무 길어서 볼 때마다 마음이 불편했는데, 그린도 그렇게 말씀해주시니 다시 한번 생각하게 되네요!

혹시 `메서드는 한 줄로 쓴다.`, `단, 일정 길이가 넘어가면 파라미터를 쉼표 단위로 개행한다`라는 규칙을 갖는 것에 대해 어떻게 생각하시나요?
(물론 개발팀에서 자체적으로 결정할 일이겠지만, 현업에서 이렇게 2개의 기준을 갖는 경우도 있는지 궁금합니다!)

GREENOVER replied:
네 맞습니다!
이러한 컨벤션을 강제적으로 swiftLint나 swiftFormat 라이브러리를 사용해 강제화 시켜주는 방법을 많이 택하긴 합니다.
한 라인에 120자를 넘어가면 워닝을 띄워주거나 하는 컨벤션을 주로 사용합니다.
그런데 메서드가 길이가 짧다라는것도 팀원 개개인이 느끼기에 매번 주관적일것이기에 만약 자유도를 조금 낮추고 모든것을 컨벤션으로 지켜주고 싶다면 그러한 부분까지도 규칙을 갖는게 적절하지 않을까란 생각이 있어요.
물론 자유도를 조금 높여서 짧은데도 강제로 줄바꿈하는것보다 한 라인에 표현하는게 더 가독성이 좋을때도 있기에 이는 각 팀에서 선택할 영역이긴합니다🙌
그래서 결국 여쭤보신 메서드는 한줄, 단 일정 길이 이상 시 파라미터를 쉼표 단위로 개행 이 부분에 있어서 적극 찬성입니다.
여기서 일정 길이를 잘 정해보면 좋을것 같아요👍

GREENOVER commented on OpenMarket/OpenMarket/Sources/Models/DataManager.swift:
```diff
@@ -0,0 +1,79 @@
+    // MARK: - Static Actions
+    
+    static func performRequestToAPI(with request: String, completion: @escaping (Result<Data, APIError>) -> Void) {
+        let requestURL = DataManager.openMarketHostAPI + request
```
추후 host가 여러개일때 이 메서드를 재사용하기가 힘들어보이는데 괜찮을까요?

yeeton37 replied:
넵. `performRequestToAPI` 메서드의 매개변수로 hostAPI, request를 받아주어 추후에 hostAPI와 나머지 request 부분을 재사용할 수 있도록 변경해주었습니다.

[변경 후 코드]
```swift
static func performRequestToAPI(from hostAPI: String, with request: String,
                                    completion: @escaping (Result<Data, NetworkingError>) -> Void) {
        let url = hostAPI + request
        
        guard let url = URL(string: url) else {
            return
        }
```

GREENOVER commented on OpenMarket/OpenMarket/Sources/Models/DataManager.swift:
```diff
@@ -0,0 +1,79 @@
+        return dataAsset.data
+    }
+    
+    static func parse<T: Decodable>(_ data: Data?, into target: T) -> T? {
+        let jsonDecoder: JSONDecoder = JSONDecoder()
+        let targetType = type(of: target)
```
target을 받아 타입을 추론하기보다 타입을 제네릭하게 받으면 어떨까요?🙋🏻
예를들어 현재 구조에서는 
`var itemListPage: ItemListPage?` 를 변수로 만들어 이를 통해 parse에 넘겨주기보다
parse에서 알아야할건 이 디코딩을 위한 모델 타입이 중요한거라 생각되어 ItemListPage 즉 이 타입 자체를 target으로 넘겨주면 어떨까 싶어요🤔
제 의견은 의견일 뿐이니 의도하신 다른 이유가 있으면 꼭 말씀해주세요🙌

yeeton37 replied:
parse의 매개변수로 타입 자체를 넘겨주는 방법이 있을 거라고 생각을 못했네요!! 너무 좋아요😁

아래와 같이 기존에 제너릭 타입(T)에서 -> 제너릭의 metaType(T.Type)을 매개변수로 받는 것으로 변경해주었습니다.
```swift
 static func parse<T: Decodable>(_ data: Data?, into type: T.Type) -> T? {
        let jsonDecoder: JSONDecoder = JSONDecoder()
        
        guard let data = data else {
            return nil
        }
        
        do {
            let decodedData = try jsonDecoder.decode(T.self, from: data)
            return decodedData
```

GREENOVER commented on OpenMarket/OpenMarket/Sources/Models/DataManager.swift:
```diff
@@ -0,0 +1,79 @@
+
+// MARK: - Private Static Actions
+
+private extension DataManager {
+    static func isValidResponse(_ response: URLResponse?) -> Bool {
+        guard let httpResponse = response as? HTTPURLResponse,
+              (200...299).contains(httpResponse.statusCode) else {
+            return false
+        }
+        
+        return true
+    }
```
LGTM👍👍👍

minsson replied:
😄

GREENOVER commented on OpenMarket/OpenMarket/Sources/Controllers/ItemListPageViewController.swift:
```diff
@@ -0,0 +1,78 @@
+
+import UIKit
+
+final class ItemListPageViewController: UIViewController {
```
명확한 VC 네이밍 좋네요👍

minsson replied:
감사합니다! 😁

GREENOVER commented on OpenMarket/OpenMarket/Sources/Controllers/ItemListPageViewController.swift:
```diff
@@ -0,0 +1,78 @@
+// MARK: - Private Enums
+
+private extension ItemListPageViewController {
+    enum QueryValue {
+        static var pageNumber = "1"
+        static var itemsPerPage = "10"
+    }
+    
+    enum QueryKey {
+        static var pageNumber = "page_no="
+        static var itemsPerPage = "items_per_page="
+    }
+    
+    enum Path {
+        static var products = "/api/products"
+    }
+    
+    enum QueryCharacter {
+        static var questionMark = "?"
+        static var ampersand = "&"
+    }
```
얘네는 충분히 다른 VC에서도 사용될 가능성을 배제할 수 가 없다고 생각하는데
그렇다면 별도 네임스페이스로 파일로 빼는건 어떨까요?🙋🏻
최대한 이러한 네임스페이스들은 하나의 그룹에서 파일로 구분하여 관리하면 추후 유지보수에 편리합니다!

yeeton37 replied:
오..좋아요! 저희는 추후에 다른 VC들을 구현할 때 중복되는 enum의 case들을 한번에 별도의 네임스페이스로 분리해주고 싶어서 step1에서는 이렇게 VC 내부에 enum을 구현해주었습니다!
step2를 진행하면서 겹치는 부분들을 별도의 네임스페이스 파일을 만들어서 빼주도록 하겠습니다 ☺️

GREENOVER commented on OpenMarket/OpenMarket/Sources/Controllers/ItemListPageViewController.swift:
```diff
@@ -0,0 +1,78 @@
+            self.itemListPage = parsedData
+            
+        case .failure(let error):
+            print(error)
```
실제 performRequestToAPI를 호출해 실패하여 이 부분으로 패치 에러로 떨어진다면 사용자에게 더 좋은 경험을 주려면 어떻게 해야할까요?
현재에서는 단순 개발자의 디버깅 그 이상이 아니라고 보여져서요..!

minsson replied:
`서버에서 데이터를 가져오는 과정에서 에러가 발생했습니다. 잠시 후 다시 시도해주세요.`와 같은 Alert를 띄워주면 좋을 것 같아요! 이 부분은 Step 2에서 UI를 구현할 예정이니, 그때 함께 수정해보겠습니다.

개인적으로 에러메시지는 제가 작성한 것처럼 딱딱한 말투보다는 좀 더 귀여운(😅) 말투가 좋을 것 같다고 생각해요. 앱이 사용자에게 대화를 건네는 문체에 따라 에러가 발생해도 부정적인 감정이 덜 들고, 사용자 경험이 좋아지더라고요. (물론 에러가 발생하지 않는 게 가장 좋겠지만요) 그래서 이 부분은 현업에 들어가게 되면 기획팀과 대화하고 결정하면 좋을 것 같습니다. 😁

그린은 어떻게 생각하시나요?

GREENOVER replied:
타당한 근거에 의한 맞는 말씀이라고 생각하며 저도 동의합니다!
사용자 친화적인 문구와 관련해서는 현업에선 당연히 기획 / 디자인 팀과 같이 소통할테니 크게 프론트팀에서는 크게 고민하실건 없을것 같고 그렇다면 프론트팀에서 고민해야하는건 얼럿을 띄워주기만할것인지 얼럿을 띄워준 후 실제 다시 재통신을 시킬것인지 고런 부분들을 고려해보면 좋을것 같습니다🙌
스텝 2에서 충분히 원하시는 방향으로 해보셔도 좋을듯해요!
필수 요구사항은 아니기에 정말 여러분들이 해보고 싶고 사용자 친화적인 방향이면 좋겠습니다~!

minsson replied:
다시 재통신을 시킨다는 부분은 미처 생각해보지 못했네요! 사용자가 Alert를 끄고, 원한다면 다시 시도하는 방향으로 생각했었어요. 

필요시에는 사용자에게 `로딩 중` 같은 Alert를 띄워놓고, 정해진 횟수만큼, 혹은 정해진 시간 동안 해당 메서드를 실행한 후, 최종적으로 실패했다는 Alert를 띄우는 방향도 고려할 수 있을 것 같아요! 이렇게 진행하면 사용자는 여러 번 실패했다는 걸 인지하지 못할 수 있겠네요 😈

이 부분은 어떻게 처리할지 고민해보겠습니다! 😁

GREENOVER commented on OpenMarket/OpenMarket/Sources/Controllers/ItemListPageViewController.swift:
```diff
@@ -0,0 +1,78 @@
+        }
+    }
+    
+    func fetchDataForItemListPage() {
+        let data = DataManager.makeDataFrom(fileName: "products")
+        
+        guard let parsedData = DataManager.parse(data, into: itemListPage) else {
+            return
+        }
+        
+        itemListPage = parsedData
+    }
```
필요한 메서드인가요~?🙋🏻
mockDate를 위한 메서드였으면 이제 제거해도 무방하지 않을까 싶습니다~!

minsson replied:
유닛테스트에서 `fetchDataForItemListPage()` 내부에서 호출하던 메서드 2개를 따로따로 호출하고 있으므로, 이제 필요가 없는 메서드입니다! 놓치고 지나갔으면 나중에 레거시 코드가 되었을 수도 있겠네요..!! 😅 제거했습니다!

GREENOVER commented on OpenMarket/OpenMarket/Sources/Controllers/ItemListPageViewController.swift:
```diff
@@ -0,0 +1,78 @@
+//
+//  OpenMarket - ViewController.swift
```
사소하지만 요런 주석의 네이밍도 같이 맞춰주면 좋을것 같아요👍

minsson replied:
그렇네요! 수정했습니다 😁

GREENOVER reviewed (changes requested):
@minsson @yeeton37 
두분 안녕하세요🙇🏻
이번 프로젝트 리뷰하게 되어 영광이며 잘 부탁드려요!ㅎㅎ
아무래도 이번 프로젝트가 처음 접하는 개념들도 많고 처음 접한다면 학습할것이 정말 정말 많았을거에요!
그럼에도 가장 어떻게보면 프론트엔드 개발자로써 가장 중요한 기초를 쌓는 프로젝트니 꼭 잘 저랑 같이 끝까지 완주해봐요🙌
두분 우선 너무너무 스텝1 짜임새있게 잘해주셨어요👍
제가 코멘트가 다소..? 많게 느껴지실수도 있지만 그만큼 잘 해주셔서 저도 궁금한 부분들과 조금 더 보완되면 완벽할것 같은 부분들에 주로 코멘트를 남겼습니다ㅎㅎ
그리고 항상 시작때 말씀을 드리는 부분인데..!
제 코멘트와 리뷰는 정답을 드리는것이 아닙니다.
그렇기에 여러분들의 의도가 더 적절하고 맞다고 판단되면 그래도 킵고잉하시며 저에게 근거만 말씀해주시면 오케이입니다!
그럼 리뷰에 대해 얘기해볼께요🙌
***
### 칭찬 드리고 싶은 부분
- 파일 구조를 잘 나눠주셨네요 💯
- 모든 부분에서 완성도가 좋아요 💯
- 네이밍이 좋아서 가독이 편리했어요 💯
- 메서드를 관심사별로 작은 단위로 잘 쪼개서 더할 나위 없네요 💯
- 리드미 최고 그자체 💯💯💯
***
### 고민하셨던 부분에 대한 저의 의견
1. 네트워크 상황과 무관한 네트워킹 데이터 타입의 Unit Test
- 이 부분은 정말 여러분들이 잘 파악하신것처럼 일장일단이 있다고 생각해요.
그런데 우선 결론을 말씀드리면 저는 무리하게 테스트코드에 맞추자고 효율적인 구조를 변경하는것을 지양합니다.
우리의 목적은 테스트가 아니잖아요?
그렇기에 꼭 테스트에 맞추려면 효율적인 구조를 변경하기보다 별도로 테스트에 맞는 타입을 정의하여 (테스트만을 위한 타입이죠) 사용하는것이 더 적절할것 같아요.
특히 parse를 통해 확인할 수 있다고 판단되시면 더더욱 그럴것 같습니다.
테스트 코드 정말 중요하지만 그것에 매몰되어 가장 중요한 본질을 망각되지 않았으면 좋겠어요🙌
2. 각 화면별 모델의 구성
- 우선 이건 각 회사마다 어떻게 서버에서 받아온 데이터를 어떤 시점에 다뤄줄지의 차이라고 생각해요.
당근마켓에서 테이블 리스트를 다 받아온 후 네트워크 단절 후 상품 정보를 접근하면 회색화면 혹은 정상적으로 보이지 않는것은 기술적으로 불가능한것이 아닌 이것까지 고려할 이유가 없는것이라 생각합니다.
즉 생각해볼께요.
우선 앱이 실행될때 상품 리스트들을 쭉 받아옵니다.
이건 실제 뷰를 모든 상품들을 다 받아온 다음 짠하고 다같이 나타내게 해줄 수도 있는것이고 페이지네이션을 통해 일정 100개 받아오고 또 100개 받아오고 이런식으로도 할 수 있겠죠.
많은 상품을 다 받아와야하는 당근마켓과 같은 대규모의 앱이라면 이러한 페이지네이션을 이용했을거라 생각해요.
그렇기에 실제 사용자는 부드러운 사용성을 얻을 수 있겠죠.
또한 이렇게 받아온 데이터들은 캐싱을 해두게 됩니다. 매번 이미지같은 경우 사이즈가 클텐데 불러오면 속도 저하나 사용성에 좋지 않으니까요.
그렇기에 예를들어 이미지 캐싱을 위한 Kingfisher, Nuke, Alamofire에도 이러한것이 있고 심지어 애플에서 제공하는 AsyncImage와 같은 기술들 정말 다양한 캐싱 기술중 하나를 택해 사용하지 않을까해요.
그렇다면 한번 통신한 후로는 앱을 재실행하지 않는한 캐싱된 메모리를 사용하겠죠.
그 다음 상품 상세정보로 넘어가봅시다!
아마도 상품 상세 정보를 클릭할때 API 호출을 할거에요. 
그리고 호출 된 후 정보에 대해서는 캐싱할거니 한번 눌렀던 상품에 대해서는 조금 더 빠른 접근이 될 수도 있겠죠.
여기서 당연히 상품을 눌러서 상품 디테일뷰로 접근할때 API 서버 통신이 발생하니 네트워크가 단절되었다면 뷰를 띄울 수 없겠죠 정보가 없으니..!
즉 이러한 데이터들을 상품 목록을 가져올때 불러와서 데이터를 가지고 있을 수 있지만 그럴 필요가 있으며 성능까지 저하시키며 가져가야할 사항인가 고려했을땐 아닐거에요.
왜냐면 예를들어 100개의 상품 목록중 이 유저는 한개만 클릭하고 나갈 수도 있는건데 모든 상품에 대한 데이터를 클릭하기도 전에 가지고 있어야 가능한 상황이 펼쳐지기에 목적에 맞지 않다고 생각해요🤔
그리고 언급하셨던 **"실제로 네트워크에 접속 중인 상태에서도 화면 이동시 물품 이미지 사진 로딩에 아주 약간의 지연이 발생하며, 로딩되는 동안 회색화면으로 보이는 것을 알 수 있었습니다."**
요 부분에 대해서는 위에서 말했던 이미지 캐싱을 통해주는데 최초 접속시에는 받아온 이미지에 대해 캐싱하기에 아주 약간의 지연이 발생할것이며 아마 동일한 상품에 대해서 다시 접근하거나 한다면 그런 지연도 전혀 발생하지 않을거에요.
이건 어느정도 매끄럽게 가져오느냐의 차이이지 지연이 아예 없을수는 없을거에요🥲
그렇기에 이미지를 패칭받는동안 대체 이미지로 회색화면을 선택했다고 볼 수 있을것입니다.
음.. 제가 주절주절 말한것 같은데 글쓰는 재주가 없어서 전달이 잘 되었는지 모르겠네요ㅠ
이해가 안가시거나 추가 설명이 필요하다고 판단되면 언제든 말씀해주세요! 음성이라든지 편한 방법으로 얘기해봐요🙌
3. json 데이터 파싱
- 전체적으로 JSON 데이터 파싱을 위한 모델 타입 설계 아주 잘해주셨어요!
하나 이해가 잘 가지 않는 질문 부분은 name 즉 상품 이름에 대해 서버에서도 정해진 enum에 따른 값을 넘겨주고 이를 클라에서도 Currency와 같이 enum으로 관리해 처리해주면 어떻겠느냐는 말씀이실까요~?🙋🏻
실제적으로 어떤 플로우를 말씀하신지 제가 잘 이해를 못했습니다🥲
그럼에도 말씀드려보면 우선 name이 정말 상품 이름이기에 무수히 많을것인데 상품 등록시 만들어진 name을 열거로 관리하기엔 음.. 어떤 효용이 있을지 잘 모르겠어요.
말씀하신 name의 오타가 어떨때 날지에 대해 감이 오지 않습니다!
오타가 날 일이 있나요..?
그 외 데이터에 관해 서버와 당연히 클라도 충분히 말할 수 있습니다.
애초에 API 설계 단 부터 같이 참여를 하며 response와 requestDTO 정의 시 목소리를 내는게 가장 중요하며 그에 따라 필드 즉 클라에서는 모델 타입 설계 시에 필요한 이런 name과 같은 부분들도 어떤 타입, 형식으로 소통할지 충분히 의논할 수 있습니다!
4. URLSessionDataTask - init() deprecated
- 음 좋습니다!
우선 질문을 드려보고 싶어요ㅎㅎ
더 좋은 방법에 대해 여쭤보시는것이 더 쉬운 방법을 뜻하는것일까요?
그게 아니라면 찾아보신 두 방법에 대해 어렵더라도 조금 천천히 이해를 해보고 이게 좋은 방법인지 판단을 먼저 해보고 다른 방법을 찾아보면 어떨까라는 생각이 있습니다!
지금 찾은 방법들에 대해 단순 어려워서 다른 방법을 찾는것은 조금 위험할것 같아요🥲
만약 다른 조금 더 쉬운 방법이 있어서 그걸 공유드린다면 여러분들은 다른 방법에 대해 학습해볼 기회를 놓친다고 생각해요.
지금은 이 방법을 쓰고 다음에 다른 방법을 공부해보자~라는 생각을 대게 하게되는데 분명 그럴 기회도 시간도 별로 없을거에요ㅠㅠ
처음 접하는것들이니 어려운건 당연합니다! 
어려운걸 만났을때 이게 맞는 방법인데 단순 어려울것 같아 다른 방법을 찾는것보다 현재 만난 이 문제를 어떻게든 이해하고 확실히 잡고 넘어간다음 더 보완된 방법을 찾아보는게 어떨까요?
결국 경고를 발생시키는 근본 원인은 모든 방법의 해결책이든 동일하기에 근본적인 원인이 왜 발생했는지에 대해 이해가 된다면 어떤 해결책이든 조금 더 쉽게 이해가 될거에요🙌
여러분들은 충분히 조금 시간을 들이면 해결할 수 있고 이해할 수 있으실거라 믿어요.
5. 구조체 타입 내부 Enum
- 이 부분은 코멘트로도 남기면서 여러분의 의견을 여쭤봤는데 어느정도 의견이 여기에 있긴했군요!ㅋㅋ
우선 내부에 있다면 당연히 private함으로 관심사가 여기에만 있다는 가정하에서는 의도대로 좋은 방법입니다.
그런데 현재도 그렇지만 nest되는 뎁스가 2개 이상 넘어가면 가독성을 해칠 경우도 있어요.
즉 현재 방법의 장점은 구현해주고 채택해주신 이유가 될것같고 굳이 단점이라 한다면 코드가 복잡해질 시 가독성과 해당 enum을 찾기 어려울 수 있다 정도로 생각합니다.
별도 파일로 네임 스페이스나 enum을 분리해준다면 가장 좋은 장점은 가독성과 유지보수라고 생각해요.
프로젝트 트리에 파일이 몇백개가 있다고 가정했을때 어떤 파일에서 이 타입안에 중첩된 타입들을 기억하고 찾아 들어가기가 어려울 수 있습니다.
그렇기에 별도로 파일로 분리한다면 조금 더 유지보수 측면에서 좋지 않을까 싶습니다.
이건 정말 답이 없는 스타일의 차이이고 여러분들이 속한 팀 향후 회사의 코드 컨벤션의 일종이라고 볼 수도 있을것 같아요!
민쏜의 의견에도 동의합니다. 예톤의 의견은 어떨지도 공유해주시면 좋을것 같습니다👍
***
### 보완되면 좋을 부분
- 너무 잘해주셔서 큰 부분은 없고 조금 더 완성도 있게 리팩토링 해보면 좋을것 같습니다!
- 현재 스텝에서 이뤄지지 않아도 되는 부분이긴 하지만 테스트 코드에 대해 조금 더 다양하면 베스트일것 같아요ㅎㅎ
***
두분 다시 한번 너무너무 고생 많으셨어요!
일주일 고생하신만큼 완성도가 깊네요👍
우선 리팩토링 해볼 부분이 있을것 같아 Request Changes로 남겨두겠습니다!
언제든 재 리뷰를 원하신다면 우측 리프레쉬 버튼 및 디엠을 통해 요청해주세요~
그럼 주말 잘 보내시고 화이팅입니다!

yeeton37 committed:
refactor: "NetworkingError" 타입명 및 case 추가 및 수정

minsson committed:
refactor: "ItemListPageViewController" 불필요한 메서드 제거

- fetchDataForItemListPage()의 경우, Mock 데이터로 유닛 테스트할 때 쓰기 위해 만든 메서드
- 실제로 유닛 테스트 시에는 이 메서드 안에 작성되어 있던 메서드들을 개별적으로 호출해 사용함
- 따라서 필요하지 않은 메서드이므로 제거

yeeton37 committed:
refactor: "HostAPI" hostAPI에 대한 네임스페이스 타입 생성 및 적용

- 추후에 여러 개의 hostAPI가 생성될 수 있음을 고려하여 별도의 네임스페이스를 생성

yeeton37 committed:
refactor: "DataManager" parse 메서드의 매개변수 타입 수정 및 적용

- 기존에는 parse의 매개변수로 인스턴스를 받아와서 메서드 내부에서 타입을 추론 후 decode 메서드 안에 넣어주었음
- 하지만 decode할 타입 자체를 매개변수로 받아올 수 있다는 것을 알게되었음
- 이에 따라 매개변수로 타입 자체를 받아줄 수 있도록 제너릭 타입(T)에서 제너릭의 metaType(T.Type)으로 변경함

yeeton37 committed:
refactor: "DataManager" url 재사용을 위해 매개변수를 추가

- 메서드 내부에서 정해져있던 url을 함수의 인자로 받아오는 것으로 변경하면서 추후에 url을 재사용할 수 있도록 함

minsson committed:
refactor: "DataManager" NetworkManager로 타입 이름 변경

- API 네트워크 통신하는 역할 및 지정된 모델 타입으로 디코딩하는 역할 모두를 수행하므로, 더 범용적인 네이밍으로 수정

minsson committed:
fix: "DataManagerTests" 오류 발생시에도 success 가능한 문제 수정

minsson committed:
refactor: "DataManagerTests" 불필요한 변수 제거

yeeton37 commented:
안녕하세요, 그린! 
위에서 답변해주신 내용에 대해 저희가 이해한 바를 아래에 정리해보았습니다. 해결되지 않았던 1,4번 고민은 디스코드를 통해 그린과 대화를 나누어서 따로 적지 않았습니다.


#### 1. 각 화면별 모델의 구성

- 저희가 고민했던 것은 당근마켓 앱을 실행하면서 화면의 테이블뷰에서 데이터를 미리 어느정도 받아놓는 것인지, 혹은 네트워킹을 통해 그때 그때 받아오는 것인지?에 대한 것이었습니다.
- 그린의 코멘트를 읽고 나서 이해한 바로는 실제 뷰를 띄울 때 모든 상품들을 다 받아온 다음 다같이 나타내게 해줄 수도 있는 것이고, 페이지네이션을 통해 조금씩 필요한 데이터를 그때그때 받아올 수도 있는데 당근마켓과 같은 대규모의 앱이라면 조금씩 필요한 데이터를 그때그때 받아오는 페이지네이션을 이용했을 것이라는 것이었습니다.
- 또한 캐싱이라는 개념에 대해 아직 정확히 공부해보진 못했지만, 기기를 끄기 전까지 임시로 저장해두는 NSCache라는 클래스를 사용하여 한번 받아온 데이터는 캐싱을 해두어 나중에 더 빠르게 불러올 수 있다는 사실도 알게되었습니다. 
- 결론적으로 그때그때 네트워킹을 통해 데이터를 불러오는 것은 그렇게 할 수는 있지만 속도 저하나 사용성에 좋지 않을 수 있고, 또 데이터를 미리 모두 받아놓는 것은 사용자가 한 개의 데이터만 보고 나가는 등의 상황을 고려했을 때 성능상 문제가 발생할 수 있다고 이해했습니다.
저희가 이해한 바가 맞는지 모르겠지만... 그린의 코멘트를 읽고 이해한 바를 위와 같이 정리해봤습니다. 

#### 2. json 데이터 파싱
- 저희가 생각한 방향이 애초에 잘못된 접근이었다는 걸 알게되었습니다.. 😅 

#### 3. 구조체 타입 내부 Enum
- 코드 작성 당시 민쏜의 생각은 클래스 타입 내부에서만 사용되는 Name Space들을 정의해주기 위해서 타입 내부에 enum을 선언해준 뒤 private으로 접근 제한을 걸어주자는 의견이었습니다. 
- 저(예톤)는 이렇게 클래스 내부에서만 프로퍼티를 사용할 때에는 enum을 클래스 내부에 선언해줄 수 있다는 사실을 처음 알았고, 원래는 항상 네임스페이스를 정의해줄 때는 전역으로 파일을 생성해서 만들어주었었습니다. 
- 그럼에도 민쏜의 의견에 따라간 이유는 그당시 프로젝트의 모든 요구사항을 알지 못했고 프로젝트에 대한 전체적인 방향성을 알지 못했던 상황에서 일단 이미 구현해놓은 뷰컨에만 해당되도록 네임스페이스를 지정해주고, 추후에 뷰컨이 더 추가되고 프로젝트의 큰그림이 잡혔을 때 모든 중복되는 타입들을 전역으로 빼주자고 하는 민쏜의 의견에 적극 동감했기 때문입니다. 😃

감사합니다 그린!🍀🍀

GREENOVER reviewed (approved):
고생 많으셨습니다!
스텝1에서 필요한 요구사항은 모두 충족하였다고 판단하고 추가로 더 리팩토링 할 부분들이 있다면 스텝2에서 같이 녹여봐요ㅎㅎ
어프로브 및 머지하겠습니다!
스텝2도 화이팅이에요🙌
